### ISSUE TITLE
removing node should not result in Karpenter panicing

### ISSUE THREAD

**Issue by BryanStenson-okta** [aws/karpenter#1703]
## Version
**Karpenter**: v0.9.0-rc0
**Kubernetes**: Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.9-eks-0d102a7", GitCommit:"eb09fc479c1b2bfcc35c47416efb36f1b9052d58", GitTreeState:"clean", BuildDate:"2022-02-17T16:36:28Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}

## Expected Behavior
removing a node should not result in Karpenter dying via panic

## Actual Behavior
```
2022-04-19T23:30:04.651Z	ERROR	controller	unable to find node for bound pod, with node name %!c(string=ip-10-41-150-171.us-west-2.compute.internal)	{"commit": "c9b3eb3"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13f923e]

goroutine 2173 [running]:
github.com/aws/karpenter/pkg/controllers/state.(*Cluster).ForPodsWithAntiAffinity.func1({0xc00292f8d8, 0x3c160f8}, {0x2348e00, 0xc000b47400})
	github.com/aws/karpenter/pkg/controllers/state/state.go:92 +0x23e
sync.(*Map).Range(0xc001187970, 0xc00292f958)
	sync/map.go:346 +0x2aa
github.com/aws/karpenter/pkg/controllers/state.(*Cluster).ForPodsWithAntiAffinity(0x21604e0, 0x1fc9ba0)
	github.com/aws/karpenter/pkg/controllers/state/state.go:80 +0x49
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).updateInverseAffinities(0xc002c3f1d0, {0x27fda30, 0xc002895e00})
	github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:184 +0xb7
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.NewTopology({0x27fda30, 0xc002895e00}, {0x283e058, 0xc00036e5f0}, 0xc0033df9d0, {0xc002bd5ac0, 0x7, 0xa})
	github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:68 +0x209
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Scheduler).Solve(0xc003c60f00, {0x27fda30, 0xc002895e00}, 0xc0033df998, {0xc002b8e000, 0x19a, 0x200}, {0xc0027f0618, 0x1, 0x1})
	github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/scheduler.go:100 +0x5a8
github.com/aws/karpenter/pkg/controllers/provisioning.(*Provisioner).provision(0xc000a23a90, {0x27fda30, 0xc002895e00})
	github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:108 +0x346
github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner.func1()
	github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:52 +0x70
created by github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner
	github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:50 +0x2dd
```

**Comment by BryanStenson-okta** [comment]
I realize this is a pretty slim bug report.  I wanted to get something filed in case some one else noticed this.  I'll try to add more details/steps if/when I get an accurate reproduction.

**Comment by tzneal** [comment]
@BryanStenson-okta  I think this is from the experimental branch that you are running, still a good bug report but I don't think this is from the RC as that cluster state tracking code isn't in v0.9.0-rc.0.

**Comment by tzneal** [comment]
Thanks for this, I wrote up a test and fixed it in #1698 .  This is a preliminary PR that starts to add support for maintaining cluster state as we move towards in-flight node awareness and potentially not-binding.

**Comment by tzneal** [comment]
Cluster state with fixes was merged.
